File: code/moments_from_coefficients/coefficients_to_micro_data.jl
# ─────────────────────────────────────────────────────────────
# coefficients_to_micro_data.jl — synthetic micro data from the
# published coefficient files.
# ─────────────────────────────────────────────────────────────
#
# WHAT THIS PRODUCES
# One weighted synthetic cross-section per quarter, in the same layout the
# model pipeline writes (see `construct_micro_dataset` in Correlations.jl and
# `create_micro_df`, called from CreateTimeSeries.jl): one row per cell of the
# 10×10×10 decile copula grid,
#
#   cop_share, grid_point, consum, income, wealth,
#   consumgrid, incomegrid, wealthgrid, time
#
#   cop_share  — probability mass of the cell (the household weight share;
#                model-implied, so isolated cells can be slightly negative)
#   grid_point — concatenated cell index, e.g. 372 = (consum decile 3,
#                income decile 7, wealth decile 2); deciles are 1 (bottom)
#                to 10 (top)
#   consum/income/wealth — the AVERAGE of the measure within its decile,
#                RELATIVE TO THE PER-HOUSEHOLD MEAN of that quarter
#   *grid      — the cell indices as separate columns
#
# HOW IT FOLLOWS THE MODEL CODE
# * Weights: `copula_pmf_grid` reproduces the published `ciw_*` probability
#   masses (the copula block of the coefficient files).
# * Values: CreateTimeSeries.jl (≈ line 2376) integrates the quantile function
#   over each decile with quadgk and divides by the interval length. Here the
#   same average is computed as the mean of `quantile_at` on 32 nodes per
#   decile (the quantile function is a degree-11 polynomial in the basis, so
#   this agrees to ~machine precision).
# * Row layout: identical to `construct_micro_dataset` (Correlations.jl),
#   including the concatenated `grid_point` index.
#
# UNITS — SCALING TO LEVELS
# Values are relative to the per-household mean (the same convention as
# `quantile_at`; e.g. 0.73 = 73% of mean consumption that quarter). The model
# pipeline multiplies by a per-household aggregate series, shipped as
# `data/synthetic/aggregate_anchors.csv` (time, consum_per_hh, income_per_hh,
# wealth_per_hh, tot_hhs — the same series the estimation uses). To get
# dollar levels, join on the quarter and compute
#
#   level_it = value_it × <measure>_per_hh_t
#
# To expand into unit-record data (N synthetic households), sample rows with
# probability `cop_share` (clip negatives to 0 and renormalize) and assign
# each household its row's values.
#
# USAGE
#   julia --project=../julia/env coefficients_to_micro_data.jl
#   # options: DATASET=PSID TREND=normal DATE=2008-Q3 (omit DATE for all quarters)
#   # output:  ../../data/synthetic/<DATASET>_synthetic_microdata.csv
#
# Only deps: CSV, DataFrames (wraps the canonical helper code/julia/reconstruct.jl).
# Python twin: coefficients_to_micro_data.py.
# ─────────────────────────────────────────────────────────────

const HERE = @__DIR__
const REPO = abspath(joinpath(HERE, "..", ".."))

include(joinpath(REPO, "code", "julia", "reconstruct.jl"))
using .DistributionalReconstruction
using CSV, DataFrames, Statistics

const MEASURES = [:consum, :income, :wealth]   # alphabetical — model convention
const GRID = 10
const NODES_PER_DECILE = 32

"Decile-average quantiles: mean of the quantile function within each decile."
function decile_averages(r, date, measure)
    avgs = zeros(GRID)
    for d in 1:GRID
        us = (d - 1) / GRID .+ (collect(1:NODES_PER_DECILE) .- 0.5) ./ (GRID * NODES_PER_DECILE)
        avgs[d] = mean(quantile_at(r, date, measure, us))
    end
    return avgs
end

"One quarter's synthetic cross-section (1000 rows) in the model's layout."
function micro_rows(r, date; digits = 6)
    pmf = copula_pmf_grid(r, date)                       # 10×10×10, matches ciw_*
    q = Dict(m => round.(decile_averages(r, date, m); digits) for m in MEASURES)
    df = DataFrame(cop_share = Float64[], grid_point = Int[],
                   consum = Float64[], income = Float64[], wealth = Float64[],
                   consumgrid = Int[], incomegrid = Int[], wealthgrid = Int[])
    for i in 1:GRID, j in 1:GRID, k in 1:GRID            # same order as construct_micro_dataset
        push!(df, (round(pmf[i, j, k]; digits),
                   parse(Int, string(i) * string(j) * string(k)),
                   q[:consum][i], q[:income][j], q[:wealth][k], i, j, k))
    end
    df[!, :time] .= date
    return df
end

function main()
    dataset = get(ENV, "DATASET", "PSID")
    trend   = get(ENV, "TREND", "normal")
    date    = get(ENV, "DATE", "")            # empty → all quarters

    csv = joinpath(REPO, "data", "synthetic", "$(dataset)_coefficients_$(trend).csv")
    isfile(csv) || error("Coefficient file not found: $csv")
    r = Reconstruction(csv)

    dates = isempty(date) ? available_dates(r) : [date]
    out = DataFrame()
    for (n, d) in enumerate(dates)
        append!(out, micro_rows(r, d))
        n % 50 == 0 && println("  $n / $(length(dates)) quarters")
    end

    dest = joinpath(REPO, "data", "synthetic", "$(dataset)_synthetic_microdata.csv")
    CSV.write(dest, out)
    println("Wrote $(nrow(out)) rows ($(length(dates)) quarters × $(GRID^3) cells) → $dest")
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: code/moments_from_coefficients/coefficients_to_micro_data.py
#!/usr/bin/env python3
"""
coefficients_to_micro_data.py — synthetic micro data from the published
coefficient files. Python twin of coefficients_to_micro_data.jl; see that
file's header for the full documentation (row layout, how it follows
CreateTimeSeries.jl / construct_micro_dataset, and units).

Short version: one weighted synthetic cross-section per quarter — one row per
cell of the 10x10x10 decile copula grid. `cop_share` is the cell's household
weight share; consum/income/wealth are decile AVERAGES relative to the
per-household mean. Scale to dollar levels with
data/synthetic/aggregate_anchors.csv (level = value * <measure>_per_hh).

    python coefficients_to_micro_data.py                        # PSID, all quarters
    python coefficients_to_micro_data.py --dataset PSID --date 2008-Q3

Dependencies: numpy, pandas.
"""
from __future__ import annotations

import argparse
import os
import sys

import numpy as np
import pandas as pd

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.abspath(os.path.join(HERE, "..", ".."))
sys.path.insert(0, os.path.join(REPO, "code", "python"))
from reconstruct import Reconstruction  # noqa: E402

MEASURES = ["consum", "income", "wealth"]   # alphabetical — model convention
GRID = 10
NODES_PER_DECILE = 32


def decile_averages(r: Reconstruction, date: str, measure: str) -> np.ndarray:
    """Mean of the quantile function within each decile (mirrors the quadgk
    decile integral in CreateTimeSeries.jl; exact for the polynomial basis)."""
    avgs = np.zeros(GRID)
    for d in range(GRID):
        us = d / GRID + (np.arange(NODES_PER_DECILE) + 0.5) / (GRID * NODES_PER_DECILE)
        avgs[d] = float(np.mean(r.quantile_at(date, measure, us)))
    return avgs


def micro_rows(r: Reconstruction, date: str, digits: int = 6) -> pd.DataFrame:
    pmf = r.copula_pmf_grid(date)                      # 10x10x10, matches ciw_*
    q = {m: np.round(decile_averages(r, date, m), digits) for m in MEASURES}
    rows = []
    for i in range(1, GRID + 1):                        # same order as construct_micro_dataset
        for j in range(1, GRID + 1):
            for k in range(1, GRID + 1):
                rows.append((round(float(pmf[i - 1, j - 1, k - 1]), digits),
                             int(f"{i}{j}{k}"),
                             q["consum"][i - 1], q["income"][j - 1], q["wealth"][k - 1],
                             i, j, k, date))
    return pd.DataFrame(rows, columns=["cop_share", "grid_point", *MEASURES,
                                       "consumgrid", "incomegrid", "wealthgrid", "time"])


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--dataset", default="PSID", choices=["PSID", "SCF", "CEX"])
    ap.add_argument("--trend", default="normal", choices=["normal", "average"])
    ap.add_argument("--date", default=None, help="one quarter, e.g. 2008-Q3; omit for all")
    args = ap.parse_args()

    csv = os.path.join(REPO, "data", "synthetic",
                       f"{args.dataset}_coefficients_{args.trend}.csv")
    if not os.path.exists(csv):
        sys.exit(f"Coefficient file not found: {csv}")
    r = Reconstruction(csv)

    dates = [args.date] if args.date else r.available_dates()
    out = pd.concat([micro_rows(r, d) for d in dates], ignore_index=True)

    dest = os.path.join(REPO, "data", "synthetic",
                        f"{args.dataset}_synthetic_microdata.csv")
    out.to_csv(dest, index=False)
    print(f"Wrote {len(out)} rows ({len(dates)} quarters x {GRID**3} cells) -> {dest}")


if __name__ == "__main__":
    main()
